### ISSUE TITLE
New in NumPy 1.14: failures in PyTables 

### ISSUE THREAD

**Issue by oleksandr-pavlyk** [numpy/numpy#10344]
In the following snippet:

```python
import numpy as np
data_proper = [(456, b'dbe', 1.2), (2, b'ded', 1.3)]
data_improper = [[456, b'dbe', 1.2], [2, b'ded', 1.3]]
descr = [('f0', '<i4'), ('f1', 'S3'), ('f2', '<f8')]
dtype = np.dtype((np.record, descr))

np.array(data_improper, dtype=dtype)
```

Execution of the last line in 1.13 used to raise:
```
TypeError: a bytes-like object is required, not 'int'
```
which allowed the logic in [numpy/core/records.py#L676](https://github.com/numpy/numpy/blob/master/numpy/core/records.py#L676) to recover and reinterpret the line as `np.array(data_proper, dtype=dtype)`.

The trouble is that in NumPy 1.14
```
ValueError: invalid literal for int() with base 10: b'dbe'
```
is being raised instead, which results in the failure. 

This is a gist of the issue behind numerous test failures in pyTables when used with NumPy 1.14.

**Comment by charris** [comment]
The error comes from cPython, but where it actually originates is hard to tell, the traceback has been truncated too far. I think we will need a bisect on this.

**Comment by eric-wieser** [comment]
Perhaps the easy fix is to just to catch `ValueError` in `rec.array` as well?

**Comment by charris** [comment]
@eric-wieser I'd worry that that might not be the only place affected. I think we need to know just what changed.

**Comment by eric-wieser** [comment]
I think the simplest reproduction is:
```python
>>> np.array(456, [('i', int)])
TypeError: a bytes-like object is required, not 'int'  # 1.13
array((456,), dtype=[('i', '<i4')])  # 1.14
```

**Comment by charris** [comment]
For reference, the 1.13.3 error is `TypeError: expected a readable buffer object`.

**Comment by charris** [comment]
I do think the catch in `records.py` is too narrow, too dependent on underlying implementation details, so that needs a fix.

**Comment by eric-wieser** [comment]
> so that needs a fix.

Frankly, I think we should deprecate that code path entirely. If you pass malformed data, expecting numpy to guess what you meant seems undesirable

**Comment by charris** [comment]
The change came in with #6053. I don't think this is so much a bug, as a change in behavior.

**Comment by ahaldane** [comment]
I'm reading along. Following on what @eric-wieser said, the change in behavior was to allow scalars to broadcast to all fields during array creation, not just during assignment. The difference in behavior of these two lines seemed inconsistent in 1.13:
```python
>>> a = np.zeros(3, dtype='i4,i4')
a[:] = 456  # no scalar assignment problem
>>> np.array([456], dtype='i4,i4')
TypeError: expected a readable buffer object  # can't initialize to scalar in 1.13
```

Furthermore, in 1.13:
```python
>>> np.array([456, 'dbe'], dtype='i4,S3')
TypeError: expected a readable buffer object
>>> np.rec.array([456, 'dbe'], dtype='i4,S3')
rec.array((456, 'dbe'), 
          dtype=[('f0', '<i4'), ('f1', 'S3')])
```
Why should the first line fail, but the second succeed?

So my feeling is that the special casing in `records.py` isn't ideal, but maybe there is away to keep things back-compatible, I'm not sure yet.

By the way, I will be at a conference later this week, so may be away from my computer.

**Comment by charris** [comment]
The problem seems isolated to the `record.py` module, so for 1.14 adding `ValueError` to the list of caught errors is probably an adequate fix. @eric-wieser's point about trying to fix up malformed input is the long term problem, and we probably need more time than 1.14 allows to deal with that.

**Comment by oleksandr-pavlyk** [comment]
@charris Thank you. The malformed input to `numpy.record.array` was all coming from hard-coded list of lists in pyTable's test files, so it is easy, albeit tedious, to modify. I am prepping a PR for pyTable now.

**Comment by oleksandr-pavlyk** [comment]
@charris I have opened two PRs for PyTables repo, one addressing errors caused by the current issues, and [the other](https://github.com/PyTables/PyTables/pull/663) dealing with deprecation warnings.

**Comment by eric-wieser** [comment]
>  so for 1.14 adding ValueError to the list of caught errors is probably an adequate fix

Agreed, although we might want to see if we can squeeze any other new errors out of the conversion too. Maybe even a catchall `Exception` would be worthwhile.

> and we probably need more time than 1.14 allows to deal with that.

Although if all our CI-testing users are already fixing this, this might be an easy time to slip in a deprecation warning too.

**Comment by charris** [comment]
OK, made the suggested fix in #10543. Note that the @ahaldane reverted the causal changes for 1.14.1, but getting a FutureWarning out there early is probably a good thing.
